extensions/amp-form/0.1/amp-form.js

```diff
@@ -347,66 +377,122 @@ export class AmpForm {
-        xhrUrl = addParamsToUrl(
-            dev().assertString(this.xhrAction_), this.getFormAsObject_());
-      } else {
-        xhrUrl = this.xhrAction_;
-        body = new FormData(this.form_);
+  /**
+   * Send a request to the form's action endpoint.
+   * @param {!Object<string, string>=} opt_extraValues
+   * @private
+   */
+  doXhr_(opt_extraValues) {
+    const isHeadOrGet = this.method_ == 'GET' || this.method_ == 'HEAD';
+    let xhrUrl, body;
+    if (isHeadOrGet) {
+      xhrUrl = addParamsToUrl(dev().assertString(this.xhrAction_),
+          this.getFormAsObject_(opt_extraValues));
+    } else {
+      xhrUrl = this.xhrAction_;
+      body = new FormData(this.form_);
+      for (const key in opt_extraValues) {
+        body.append(key, opt_extraValues[key]);
       }
-      this.actions_.trigger(this.form_, 'submit', /*event*/ null);
-      return this.xhr_.fetch(dev().assertString(xhrUrl), {
-        body,
-        method: this.method_,
-        credentials: 'include',
-        headers: {
-          Accept: 'application/json',
-        },
-      }).then(response => {
-        return response.json().then(json => {
-          this.triggerAction_(/* success */ true, json);
-          this.analyticsEvent_('amp-form-submit-success');
-          this.setState_(FormState_.SUBMIT_SUCCESS);
-          this.renderTemplate_(json || {});
-          this.maybeHandleRedirect_(
-              /** @type {../../../src/service/xhr-impl.FetchResponse} */ (
-                  response));
-        }, error => {
-          rethrowAsync('Failed to parse response JSON:', error);
-        });
-      }, error => {
-        this.triggerAction_(
-            /* success */ false, error ? error.responseJson : null);
-        this.analyticsEvent_('amp-form-submit-error');
-        this.setState_(FormState_.SUBMIT_ERROR);
-        this.renderTemplate_(error.responseJson || {});
-        this.maybeHandleRedirect_(
-            /** @type {../../../src/service/xhr-impl.FetchResponse} */ (
-                error));
-        rethrowAsync('Form submission failed:', error);
-      });
-    });
-    if (getMode().test) {
-      this.xhrSubmitPromise_ = p;
     }
+    return this.xhr_.fetch(dev().assertString(xhrUrl), {
+      body,
+      method: this.method_,
+      credentials: 'include',
+      headers: {
+        Accept: 'application/json',
+      },
+    });
+  }
+
+  /**
+   * Transition the form to the submit success state.
+   * @param {../../../src/service/xhr-impl.FetchResponse} response
+   * @return {!Promise}
+   * @private
+   */
+  handleXhrSubmitSuccess_(response) {
+    return response.json().then(json => {
+      this.triggerAction_(/* success */ true, json);
+      this.analyticsEvent_('amp-form-submit-success');
+      this.setState_(FormState_.SUBMIT_SUCCESS);
+      this.renderTemplate_(json || {});
+      this.maybeHandleRedirect_(response);
+    }, error => {
+      rethrowAsync('Failed to parse response JSON:', error);
```

Comment:
Thanks. Can you help me understand when something should be a user error? I haven't been able to find anything in the docs/comments.